File: 2127_maximum_employees_to_be_invited_to_a_meeting.cpp
class Solution {
public:
    int maximumInvitations(vector<int>& F) {
        int n = F.size(), z = 0, t = 0;
        vector<int> x(n, 0), y(n, 0);
        vector<bool> v(n, false);
        for (int f : F) x[f]++;
        queue<int> q;
        for (int i = 0; i < n; i++) if (!x[i]) q.push(i);
        while (!q.empty()) {
            int a = q.front(); q.pop();
            v[a] = true;
            int b = F[a];
            y[b] = y[a] + 1;
            if (--x[b] == 0) q.push(b);
        }
        for (int i = 0; i < n; i++) if (!v[i]) {
            int c = 0, j = i;
            while (!v[j]) {
                v[j] = true;
                j = F[j];
                c++;
            }
            if (c == 2) t += 2 + y[i] + y[F[i]];
            else z = max(z, c);
        }
        return max(z, t);
    }
};

File: 2127_maximum_employees_to_be_invited_to_a_meeting.py
class Solution(object):
    def maximumInvitations(self, fav):
        n = len(fav)
        deg = [0] * n
        dist = [0] * n
        visited = [False] * n
        for f in fav:
            deg[f] += 1
        queue = [i for i in range(n) if deg[i] == 0]
        for i in queue:
            visited[i] = True
            dist[fav[i]] = dist[i] + 1
            deg[fav[i]] -= 1
            if deg[fav[i]] == 0:
                queue.append(fav[i])
        cycles, pairs = 0, 0
        for i in range(n):
            if not visited[i]:
                count, x = 0, i
                while not visited[x]:
                    visited[x] = True
                    x = fav[x]
                    count += 1
                if count == 2:
                    pairs += 2 + dist[i] + dist[fav[i]]
                else:
                    cycles = max(cycles, count)
        return max(cycles, pairs)
